File: Common/ConsoleListChoiceReader.vb
Namespace Common

    Public Class ConsoleListChoiceReader(Of T)

        Public Sub New(ByVal List As IEnumerable(Of T))
            Me.New(List, Function(O) O.ToString())
        End Sub

        Public Sub New(ByVal List As IEnumerable(Of T), ByVal Selector As Func(Of T, String))
            Me.List = List.AsCachedList()
            Me.Selector = Selector
        End Sub

        Public Function ReadChoice() As T
            Dim Choices = New List(Of T)()

            Do
                Choices.Clear()
                For I = 0 To PageLength - 1
                    Dim Tmp As T = Nothing
                    If Not Me.List.TryGetValue(Me.ChoiceOffset + I, Tmp) Then
                        Exit For
                    End If
                    Choices.Add(Tmp)
                Next

                Dim PrevPagePossible = Me.ChoiceOffset - PageLength >= 0
                Dim NextPagePossible = Me.List.TryGetValue(Me.ChoiceOffset + PageLength, Nothing)

                Console.WriteLine()
                Console.WriteLine()
                For I = 0 To Choices.Count - 1
                    ConsoleUtilities.WriteColored($"{I + 1,4} : {Me.Selector.Invoke(Choices.Item(I))}", ConsoleColor.Cyan)
                    Console.WriteLine()
                Next
                Console.WriteLine()
                Dim S = "<- Previous page  |  Page {0:D2}  |  Next page ->  |  (Q)uit".Split("|"c)
                If Not PrevPagePossible Then
                    S(0) = New String(" "c, S(0).Length)
                End If
                S(1) = String.Format(S(1), Me.ChoiceOffset \ PageLength + 1)
                If Not NextPagePossible Then
                    S(2) = New String(" "c, S(2).Length)
                End If
                ConsoleUtilities.WriteColored(S.Aggregate(Function(A, B) A & "|" & B))
                Console.WriteLine()
                Console.WriteLine()
                ConsoleUtilities.WriteColored("Select your choice:", ConsoleColor.Green)

                Do
                    Dim Key = Console.ReadKey(True).Key

                    If Key = ConsoleKey.LeftArrow And PrevPagePossible Then
                        ConsoleUtilities.WriteColored(" <-")
                        Console.WriteLine()
                        Me.ChoiceOffset -= PageLength
                        Exit Do
                    End If
                    If Key = ConsoleKey.RightArrow And NextPagePossible Then
                        ConsoleUtilities.WriteColored(" ->")
                        Console.WriteLine()
                        Me.ChoiceOffset += PageLength
                        Exit Do
                    End If

                    If Key = ConsoleKey.Q Then
                        ConsoleUtilities.WriteColored(" Q")
                        Console.WriteLine()
                        Return Nothing
                    End If

                    Dim N = 0
                    If ConsoleKey.D1 <= Key And Key <= ConsoleKey.D9 Then
                        N = Key - ConsoleKey.D0
                    End If
                    If ConsoleKey.NumPad1 <= Key And Key <= ConsoleKey.NumPad9 Then
                        N = Key - ConsoleKey.NumPad0
                    End If

                    If 1 <= N And N <= Choices.Count Then
                        ConsoleUtilities.WriteColored($" {N.ToStringInv()}")
                        Console.WriteLine()

                        Return Choices.Item(N - 1)
                    End If
                Loop
            Loop
        End Function

        Const PageLength = 9

        Private ChoiceOffset As Integer = 0

        Private ReadOnly List As EnumerableCacher(Of T)
        Private ReadOnly Selector As Func(Of T, String)

    End Class

End Namespace

File: Common/ConsoleListChoiceReader.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ks.Common
{
    public class ConsoleListChoiceReader<T>
    {
        public ConsoleListChoiceReader(IEnumerable<T> List) : this(List, O => O.ToString())
        {
        }

        public ConsoleListChoiceReader(IEnumerable<T> List, Func<T, string> Selector)
        {
            this.List = List.AsCachedList();
            this.Selector = Selector;
        }

        public T ReadChoice()
        {
            var Choices = new List<T>();

            while (true)
            {
                Choices.Clear();
                for (var I = 0; I < PageLength; I++)
                {
                    if (!this.List.TryGetValue(this.ChoiceOffset + I, out var Tmp))
                    {
                        break;
                    }

                    Choices.Add(Tmp);
                }

                var PrevPagePossible = (this.ChoiceOffset - PageLength) >= 0;
                var NextPagePossible = this.List.TryGetValue(this.ChoiceOffset + PageLength, out var argValue);
                Console.WriteLine();
                Console.WriteLine();
                for (var I = 0; I < Choices.Count; I++)
                {
                    ConsoleUtilities.WriteColored($"{I + 1,4} : {this.Selector.Invoke(Choices[I])}", ConsoleColor.Cyan);
                    Console.WriteLine();
                }

                Console.WriteLine();
                var S = "<- Previous page  |  Page {0:D2}  |  Next page ->  |  (Q)uit".Split('|');
                if (!PrevPagePossible)
                {
                    S[0] = new string(' ', S[0].Length);
                }

                S[1] = string.Format(S[1], (this.ChoiceOffset / PageLength) + 1);
                if (!NextPagePossible)
                {
                    S[2] = new string(' ', S[2].Length);
                }

                ConsoleUtilities.WriteColored(S.Aggregate((A, B) => A + "|" + B));
                Console.WriteLine();
                Console.WriteLine();
                ConsoleUtilities.WriteColored("Select your choice:", ConsoleColor.Green);
                while (true)
                {
                    var Key = Console.ReadKey(true).Key;
                    if ((Key == ConsoleKey.LeftArrow) & PrevPagePossible)
                    {
                        ConsoleUtilities.WriteColored(" <-");
                        Console.WriteLine();
                        this.ChoiceOffset -= PageLength;
                        break;
                    }

                    if ((Key == ConsoleKey.RightArrow) & NextPagePossible)
                    {
                        ConsoleUtilities.WriteColored(" ->");
                        Console.WriteLine();
                        this.ChoiceOffset += PageLength;
                        break;
                    }

                    if (Key == ConsoleKey.Q)
                    {
                        ConsoleUtilities.WriteColored(" Q");
                        Console.WriteLine();
                        return default;
                    }

                    var N = 0;
                    if ((ConsoleKey.D1 <= Key) & (Key <= ConsoleKey.D9))
                    {
                        N = Key - ConsoleKey.D0;
                    }

                    if ((ConsoleKey.NumPad1 <= Key) & (Key <= ConsoleKey.NumPad9))
                    {
                        N = Key - ConsoleKey.NumPad0;
                    }

                    if ((1 <= N) & (N <= Choices.Count))
                    {
                        ConsoleUtilities.WriteColored($" {N.ToStringInv()}");
                        Console.WriteLine();
                        return Choices[N - 1];
                    }
                }
            }
        }

        private const int PageLength = 9;

        private int ChoiceOffset = 0;

        private readonly EnumerableCacher<T> List;
        private readonly Func<T, string> Selector;
    }
}
